private/buf/bufsync/syncer.go

```diff
@@ -629,6 +653,142 @@ func (s *syncer) readModuleAt(
+				return nil
+			}
+			// For each older commit we travel, we need to make sure it's a valid module with the expected
+			// module identity, or that the error handler would have chosen to override it.
+			var shouldAttachTagsForThisCommit bool
+			if _, readErr := s.readModuleAt(
+				ctx, branch, oldCommit, moduleDir,
+				readModuleAtWithExpectedModuleIdentity(targetModuleIdentity.IdentityString()),
+			); readErr == nil || s.errorHandler.HandleReadModuleError(readErr) == LookbackDecisionCodeOverride {
+				shouldAttachTagsForThisCommit = true
+			}
+			if !shouldAttachTagsForThisCommit {
+				// not a valid module, tags in this commit should not be attached to this module.
+				return nil
+			}
+			logger := s.logger.With(
+				zap.String("branch", branch),
+				zap.String("commit", oldCommit.Hash().Hex()),
+				zap.String("module directory", moduleDir),
+				zap.String("module identity", targetModuleIdentity.IdentityString()),
+				zap.String("module directory git start point", startPoint.Hex()),
+				zap.Strings("tags", tagsToAttach),
+			)
+			// Valid module in this commit to attach tags to. If attaching the older tags fails, we'll
+			// WARN+continue to not block actual pending commits to sync in this run.
+			bsrCommitName, err := s.oldTagsAttacher(ctx, targetModuleIdentity, oldCommit.Hash(), oldCommit.Author(), oldCommit.Committer(), tagsToAttach)
+			if err != nil {
+				logger.Warn("attach older tags failed", zap.Error(err))
+				return nil
+			}
+			logger.Debug("older tags attached", zap.String("BSR commit", bsrCommitName))
+			return nil
+		}
+		if err := s.repo.ForEachCommit(
+			forEachOldCommitFunc,
+			git.ForEachCommitWithHashStartPoint(startPoint.Hex()),
+		); err != nil && !errors.Is(err, stopLoopErr) {
+			return fmt.Errorf("looking back the start sync point: %w", err)
+		}
+	}
+	return nil
+}
+
+// lookbackStartingPoints returns git starting points for all modules directories to be synced in a
+// branch. It calculates it from the list of syncable commits, returning its earliest appearance in
+// the array. If no appearance is found, it returns the HEAD commit of the branch.
+//
+// e.g.: For branch "foo" we need to sync mods [mod1, mod2, mod3, mod4], and these are the syncable
+// commits:
+// - commit: a,        modules: mod1
+// - commit: b,        modules: mod1, mod2
+// - commit: c,        modules: mod1, mod3
+// - commit: d (HEAD), modules: mod1, mod2
+//
+// The returned starting points are:
+// - mod1: commit a (first appearance)
+// - mod2: commit b (first appearance)
+// - mod3: commit c (first appearance)
+// - mod4: commit d (HEAD, did not appear)
+func (s *syncer) lookbackStartingPoints(ctx context.Context, branch string, syncableCommits []*syncableCommit) (map[string]git.Hash, error) {
+	branchModulesForSync, ok := s.branchesToModulesForSync[branch]
+	if !ok || len(branchModulesForSync) == 0 {
+		// branch should not be synced, or no modules to sync in that branch
+		return nil, nil
+	}
```

Comment:
We already do this check one layer above, so we should not need to do this again?

Reply:
https://github.com/bufbuild/buf/pull/2374/files#r1294065870